src/azstoragetorch/io.py

```diff
@@ -0,0 +1,188 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See LICENSE in the project root for
+# license information.
+# --------------------------------------------------------------------------
+
+import io
+import os
+from typing import get_args, Optional, Union, Literal
+import urllib.parse
+
+from azure.identity import DefaultAzureCredential
+from azure.core.credentials import (
+    AzureSasCredential,
+    TokenCredential,
+)
+
+
+_SUPPORTED_MODES = Literal["rb"]
+_SDK_CREDENTIAL_TYPE = Optional[
+    Union[
+        AzureSasCredential,
+        TokenCredential,
+    ]
+]
+_AZSTORAGETORCH_CREDENTIAL_TYPE = Union[_SDK_CREDENTIAL_TYPE, Literal[False]]
+
+
+class BlobIO(io.IOBase):
+    def __init__(
+        self,
+        blob_url: str,
+        mode: _SUPPORTED_MODES,
+        *,
+        credential: _AZSTORAGETORCH_CREDENTIAL_TYPE = None,
+    ):
+        self._blob_url = blob_url
+        self._validate_mode(mode)
+        self._mode = mode
+        self._sdk_credential = self._get_sdk_credential(blob_url, credential)
+
+        self._blob_downloader = self._get_blob_downloader()
+        self._position = 0
+        self._closed = False
+
+    def close(self) -> None:
+        self._close_blob_downloader()
+        self._closed = True
+
+    @property
+    def closed(self) -> bool:
+        return self._closed
+
+    def fileno(self) -> int:
+        raise OSError("BlobIO object has no fileno")
+
+    def flush(self) -> None:
+        self._validate_not_closed()
+
+    def read(self, size: Optional[int] = -1, /) -> bytes:
+        if size is not None:
+            self._validate_is_integer("size", size)
+            self._validate_min("size", size, -1)
+        self._validate_not_closed()
+        return self._read(size)
+
+    def readable(self) -> bool:
+        self._validate_not_closed()
```

Comment:
Is it typical for `readable` to raise if the IO is closed? I feel like it should always just return whether this type is readable or not but then again, I've never tried this on a standard IO.

Reply:
Before sending out the PR, I did a pass on how closely `BlobIO` behaves to other standard library file-like objects. There was definitely some unexpected behavior and this was one of them. Specifically, whether it throws an exception is actually based on the mode that is open. Below shows some things I tried out:
```python
# Usage on writable file
>>> with open('hello.txt', 'wb') as f:
...     pass
>>> f.writable()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: I/O operation on closed file
>>> f.readable()
False

# Example usage on readable
>>> with open('hello.txt', 'rb') as f:
...     pass
>>> f.writable()
False
>>> f.readable()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: I/O operation on closed file

# Example usage on something that is readable and writable
>>> import io
>>> f = io.BytesIO(b'foo')
>>> f.close()
>>> f.readable()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: I/O operation on closed file.
>>> f.writable()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: I/O operation on closed file.
```
Specifically, it appears that it only throws an error if the file was closed and the `readable()`/`writable()` would have returned `True`. Thinking about it more, it seems like the standard-library is taking the perspective that when calling these methods, it is with the intention that the user will call one of the followup methods (e.g. `read()` or `write()`) and thus raise the error sooner to avoid a bad state/unintended logic being triggered because the file is closed.

I think it makes sense to keep parity with this behavior. Though the way it is implemented in `readable()` will not be correct once `wb` is implemented. So happy future-proof the `readable()` method as part of this PR to avoid forgetting to update it.

Reply:
I pushed up an update to `readable()` to make this a little more futureproofed as I suggested when we add support for `wb`.